flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/math/distributed/DistributedRowMatrix.scala

```diff
@@ -0,0 +1,166 @@
+                           val numRows: Int,
+                           val numCols: Int )
+    extends DistributedMatrix {
+
+
+
+  /**
+    * Collects the data in the form of a sequence of coordinates associated with their values.
+    * @return
+    */
+  def toCOO: Seq[(Int, Int, Double)] = {
+
+    val localRows = data.collect()
+
+    for (IndexedRow(rowIndex, vector) <- localRows;
+         (columnIndex, value) <- vector) yield (rowIndex, columnIndex, value)
+  }
+
+  /**
+    * Collects the data in the form of a SparseMatrix
+    * @return
+    */
+  def toLocalSparseMatrix: SparseMatrix = {
+    val localMatrix =
+      SparseMatrix.fromCOO(this.numRows, this.numCols, this.toCOO)
+    require(localMatrix.numRows == this.numRows)
+    require(localMatrix.numCols == this.numCols)
+    localMatrix
+  }
+
+  //TODO: convert to dense representation on the distributed matrix and collect it afterward
+  def toLocalDenseMatrix: DenseMatrix = this.toLocalSparseMatrix.toDenseMatrix
+
+  /**
+    * Apply a high-order function to couple of rows
+    * @param fun
+    * @param other
+    * @return
+    */
+  def byRowOperation(fun: (Vector, Vector) => Vector,
+                     other: DistributedRowMatrix): DistributedRowMatrix = {
+    val otherData = other.data
+    require(this.numCols == other.numCols)
+    require(this.numRows == other.numRows)
+
+    val result = this.data
+      .fullOuterJoin(otherData)
+      .where("rowIndex")
+      .equalTo("rowIndex")(
+          (left: IndexedRow, right: IndexedRow) => {
+            val row1 = Option(left) match {
+              case Some(row: IndexedRow) => row
+              case None =>
+                IndexedRow(
+                    right.rowIndex,
+                    SparseVector.fromCOO(right.values.size, List((0, 0.0))))
+            }
+            val row2 = Option(right) match {
+              case Some(row: IndexedRow) => row
+              case None =>
+                IndexedRow(
+                    left.rowIndex,
+                    SparseVector.fromCOO(left.values.size, List((0, 0.0))))
+            }
+            IndexedRow(row1.rowIndex, fun(row1.values, row2.values))
+          }
+      )
+    new DistributedRowMatrix(result, numRows, numCols)
+  }
+
+  /**
+    * Add the matrix to another matrix.
+    * @param other
+    * @return
+    */
+  def sum(other: DistributedRowMatrix): DistributedRowMatrix = {
+    val sumFunction: (Vector, Vector) => Vector = (x: Vector, y: Vector) =>
+      (x.asBreeze + y.asBreeze).fromBreeze
+    this.byRowOperation(sumFunction, other)
+  }
+
+  /**
+    * Subtracts another matrix.
+    * @param other
+    * @return
+    */
+  def subtract(other: DistributedRowMatrix): DistributedRowMatrix = {
+    val subFunction: (Vector, Vector) => Vector = (x: Vector, y: Vector) =>
+      (x.asBreeze - y.asBreeze).fromBreeze
+    this.byRowOperation(subFunction, other)
+  }
+}
+
+object DistributedRowMatrix {
+
+  type MatrixRowIndex = Int
```

Comment:
yup, I tried to stay general but I'm not sure I've been disciplined in using MatrixRowIndex instead of Int so maybe I should remove it and leave to an eventual refactoring the generalization.